Makefile

```diff
@@ -122,6 +122,11 @@ build/llvm.BUILT:
 		llvm-config
 	touch build/llvm.BUILT
 
+build/wasm-component-ld.BUILT: build/llvm.BUILT
+	cargo install wasm-component-ld@0.1.5 --root $(BUILD_PREFIX)
```

Comment:
Most of the other dependencies are expected to be install before running the Makefile (e.g. in the github actions workflow).   Are we sure we want to diverge for this case?

Reply:
Sorry I'm not quite sure what you mean, can you clarify? Are you asking whether to submodule/vendor the source of `wasm-component-ld` in this repository? Or otherwise explicitly check that Cargo is itself installed? Or something else perhaps?

I'll also note that `wasm-component-ld` itself has transitive dependencies, so a git submodule wouldn't vendor all the source here. It's possible to vendor all the source, but that seemed pretty heavyweight for just this.

Reply:
Sorry, maybe I wasn't clear.

My point was that the status quo that we don't currently download or install tools when running `make`.    The `Makefile` assumes that all the needed tools already exist in the PATH.

So maybe this `cargo install` command could/should go in `.github/workflows/main.yml` where we take care of pulling in other dependencies like wasmtime and clang?

Reply:
I'm certainly not suggesting we need to build the linker from source.

Reply:
To make sure I understand, you're thinking that this rule would do something like:

```
cp `which wasm-component-ld` $(BUILD_PREFIX)/bin
```

If so I can certainly add that, but that means that users building their own wasi-sdk locally will possibly have a different version than what CI in this repo builds.

Reply:
Oh I see, I hadn't quick grokked that we are directly shipping a binary to user here.

But yes, I guess that is what I mean.  However, on reconsideration it does seems a little odd that we are shipping pre-built binaries as part of the SDK.

@sunfishcode WDYT?   Given that we are building llvm itself from source should we maybe also be building this binary from source too?   Conversely, if we decide its reasonable to simply ship upstream binaries maybe we should just ship clang/llvm from the LLVM releases rather than spending all the time to build it from source?   Especially given that we always just pin to llvm releases?

Reply:
Looking at the [latest release of LLVM](https://github.com/llvm/llvm-project/releases/tag/llvmorg-18.1.2) it says:

> We suggest you use the binaries from your distribution or build your own if you rely on a specific platform or configuration.

It also looks like there's no release binaries for x64 linux or macOS at all?

Even if there were precompiled binaries, I think that we'd still have the problem of the sysroot not being pre-configured nor the default target being set.

Although don't get me wrong, if LLVM could avoid being built that would make CI and contributing here much less painful

Reply:
I agree with @alexcrichton, switching to official LLVM builds is tempting, but they very often don't have binaries for even all the major platforms. And it is convenient to have builds with a default sysroot configured in.

As I understand it, the PR here isn't shipping a pre-built binary to users. `cargo install wasm-component-ld@0.1.5` builds from source. The source comes from crates.io, rather than this repo or a submodule, so we might want to see if we can make it easier for people to build their own wasm-component-ld from source, but overall the approach seems ok to me.

Reply:
Ah I see, I didn't realize this was a build from source.

Can we add a comment for the non-rust folks like me?  Something like "Use cargo to build wasm-component-ld binary from rust source code".

Reply:
Certainly!